docs/internal/overview.md

```diff
@@ -0,0 +1,80 @@
+
+Our `language-server` can roughly be splitted into four areas:
+
+-   CSS: Provides intellisense for the things inside `<style>`. Internally it mostly forwards stuff to the `vscode-css-languageservice`.
+-   HTML: Provides intellisense for basic HTML tags like `div`, `a` etc. Internally it mostly forwards stuff to the `vscode-html-languageservice`.
+-   Svelte: Provides the diagnostics of the Svelte compiler. If you use preprocessors, those are invoked beforehand - that's why we need the `svelte.config.js` to know how to preprocess your files. It also does the formatting through `prettier-plugin-svelte` and other cross-cutting concerns like the "Extract Component" refactoring.
+-   TypeScript/JavaScript: Provides intellisense for all JS/TS related stuff. This not only includes code inside `<script>` but also inside the template the moment you use any of the Svelte specifics like bindings or template syntax.
+
+The last area, TS/JS, is where most of the work is done. Svelte is a mix of JavaScript/TypeScript inside `script` and more of it within Svelte's template syntax. To get a holistic view of the file, we need to account for both.
+This is also the reason why preprocessors such as `svelte-preprocess` cannot do type checking because it produces wrong diagnostics. To preprocess the script content, it only gets the content from the script. Think of this situation:
+
+```html
+<script lang="ts">
+    let a: number = 1;
+</script>
+{a}
+```
+
+`svelte-preprocess` only gets `let a = 1` and the diagnostics would output that `a` is never used, which is wrong.
+
+To get that holistic view, we need `svelte2tsx`.
+
+#### svelte2tsx
+
+To get a holistic view of Svelte files, we have two options:
+
+1. Write a language service ourselves which is capable of doing things like auto complete, go to definition, rename etc.
+2. Convert the Svelte code to something an existing language service can process which then does auto complete, go to defintion, rename etc for us.
+
+We chose the second option because TypeScript provides a language service which can do all the heavy lifting for us: We give it some files and then invoke methods like `getQuickInfoAtPosition` for hover info. These files need to be in a format the language service can understand: A form of JavaScript or TypeScript. `svelte2tsx` is the package which does this transformation: Pass in Svelte code and get back JSX or TSX code, depending on whether or not you use TS in Svelte. This code is not runnable in any way and does not actually exist at runtime when you run your app, it purely exists for the intellisense. The code also returns source mappings so we know which position in the original code corresponds to which generated position.
+
+#### Integration of svelte2tsx into the language-server
+
+This example shows how our `language-server` uses `svelte2tsx`:
+
+Svelte file changes:
+
+1. Svelte file comes in
+2. Transform Svelte to TSX/JSX
+3. Pass that result to the TypeScript language service
```

Comment:
Would be nice to see a sample output of the TSX to get a picture of what magic is going on behind the scenes (e.g. generation of interfaces with reserved names) to make the Language Server understand Svelte. Would particularly help me understand how to contribute to the discussion on how to expose props and events typings from Svelte components.

Reply:
Added a sample output and added some short mention of the ambient types. More details would go into a separate document purely for `svelte2tsx` - some day 😄